File: common/src/test/java/com/taskadapter/model/GTaskTest.java
package com.taskadapter.model;

import org.junit.Test;

import java.util.Collections;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class GTaskTest {

    @Test
    public void newTaskHasNoChildren() {
        GTask task = new GTask();
        assertTrue("a new task must have no children", task.getChildren().isEmpty());
        assertFalse("a new task must have no children", task.hasChildren());
    }

    @Test
    public void nullChildrenReturnsFalse() {
        GTask task = new GTask();
        task.setChildren(null);
        assertFalse(task.hasChildren());
    }

    @Test
    public void emptyChildrenListReturnsFalse() {
        GTask task = new GTask();
        task.setChildren(Collections.<GTask>emptyList());
        assertFalse(task.hasChildren());
    }

    @Test
    public void hasChildrenReturnsTrueWithChildren() {
        GTask task = new GTask();
        GTask child1 = new GTask();
        child1.setId(1011L);
        task.getChildren().add(child1);

        assertTrue(task.hasChildren());
    }

    @Test
    public void constructorFieldsAreDeepCloned() {
        var field = Field.apply("str");
        var task = new GTask();
        task.setValue(field, "123");
        var cloned = GTask.shallowClone(task);
        task.setValue(field, "updated");
        assertThat(cloned.getValue(field)).isEqualTo("123");
    }

    @Test
    public void constructorSetsEmptyCollections() {
        var task = new GTask();
        assertThat(task.getChildren()).isEmpty();
        assertThat(task.getRelations()).isEmpty();
    }

    @Test
    public void identityNullIdConvertedToZeroInIdentity() {
        assertThat(new GTask().getIdentity().getId()).isEqualTo(0);
    }

    @Test
    public void identityNullIdStaysNull() {
        var task = new GTask();
        task.setId(null);
        assertThat(task.getId()).isNull();
    }
}

File: common/src/test/scala/com/taskadapter/model/GTaskTest.scala
package com.taskadapter.model

import java.util.Collections

import org.junit.Assert.{assertFalse, assertTrue}
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.{FunSpec, Matchers}

@RunWith(classOf[JUnitRunner])
class GTaskTest extends FunSpec with Matchers {

  val field = Field("str")

  it("newTaskHasNoChildren") {
    val task = new GTask
    assertTrue("a new task must have no children", task.getChildren.isEmpty)
    assertFalse("a new task must have no children", task.hasChildren)
  }

  it("nullChildrenReturnsFalse") {
    val task = new GTask
    task.setChildren(null)
    assertFalse(task.hasChildren)
  }

  it("emptyChildrenListReturnsFalse") {
    val task = new GTask
    task.setChildren(Collections.emptyList[GTask])
    assertFalse(task.hasChildren)
  }

  it("hasChildrenReturnsTrueWithChildren") {
    val task = new GTask
    val child1 = new GTask
    child1.setId(1011l)
    task.getChildren.add(child1)
    assertTrue(task.hasChildren)
  }

  describe("constructor") {
    it("fields are deep cloned") {
      val task = new GTask()
      task.setValue(field, "123")
      val cloned = new GTask(task)
      task.setValue(field, "updated")
      cloned.getValue(field) shouldBe "123"
    }

    it("sets empty collections") {
      val task = new GTask()
      task.getChildren shouldBe empty
      task.getRelations shouldBe empty
    }
  }

  describe("identity") {
    it("null Id converted to 0 in Identity") {
      new GTask().getIdentity.getId shouldBe 0
    }

    it("null Id stays null") {
      val task = new GTask()
      task.setId(null)
      task.getId shouldBe null
    }
  }
}